File: cses/03_dynamic_programming/2181_counting-tilings/solution.py
# Counting Tilings - https://cses.fi/problemset/task/2181
# Đếm số cách lát kín lưới n x m bằng các quân domino 1x2 và 2x1, lấy modulo 1e9+7.
# Phương pháp: profile DP theo từng cột, bitmask trên n hàng (n <= 10 nên 2^n <= 1024).
# Tối ưu cho CPython: sinh sẵn một hàm cộng dồn (code-gen) nhằm đưa toàn bộ vòng lặp
# cộng bên trong xuống bytecode, tránh chi phí vòng lặp Python cấp cao.
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    m = int(data[1])
    MOD = 10 ** 9 + 7

    # Số cách lát lưới (n x m) bằng số cách lát (m x n) do đối xứng chuyển vị
    # (domino ngang đổi thành dọc). Chọn n là chiều nhỏ hơn để 2^n nhỏ nhất.
    if n > m:
        n, m = m, n

    size = 1 << n  # số trạng thái bitmask trên n hàng

    # incoming[nxt] = danh sách các mask cur có thể sinh ra nxt.
    # cur = tập các ô của cột hiện tại đã bị một quân domino ngang từ cột trước "thò"
    # sang chiếm chỗ. Ta lát đầy phần còn lại của cột bằng domino dọc (trong cột)
    # và domino ngang (thò sang cột kế); nxt = tập các ô bị thò sang cột kế tiếp.
    incoming = [[] for _ in range(size)]
    for cur in range(size):
        # duyệt DFS lặp: điền các ô từ hàng 0 đến hàng n-1.
        stack = [(0, 0)]  # (hàng đang xét, mask các ô thò sang cột kế)
        while stack:
            row, nxt = stack.pop()
            if row == n:
                incoming[nxt].append(cur)
                continue
            if (cur >> row) & 1:
                # ô này đã bị chiếm từ cột trước -> bỏ qua, không thò sang cột kế.
                stack.append((row + 1, nxt))
            else:
                # lựa chọn 1: đặt domino ngang thò sang cột kế -> bật bit row trong nxt.
                stack.append((row + 1, nxt | (1 << row)))
                # lựa chọn 2: đặt domino dọc phủ (row, row+1) ngay trong cột này.
                if row + 1 < n and not ((cur >> (row + 1)) & 1):
                    stack.append((row + 2, nxt))

    # Sinh sẵn hàm step(dp) -> ndp: mỗi phần tử ndp[nxt] là tổng đã biết trước của
    # các dp[cur] (cur thuộc incoming[nxt]), rút gọn modulo. Nhờ hardcode các chỉ số,
    # phép cộng dồn chạy trực tiếp ở tầng bytecode nên rất nhanh trên CPython.
    parts = []
    for nxt in range(size):
        lst = incoming[nxt]
        if lst:
            parts.append("(" + "+".join("d%d" % c for c in lst) + ")%" + str(MOD))
        else:
            parts.append("0")
    src = (
        "def step(dp):\n"
        "    " + ",".join("d%d" % i for i in range(size)) + ",=dp\n"
        "    return [" + ",".join(parts) + "]\n"
    )
    ns = {}
    exec(src, ns)
    step = ns["step"]

    # DP qua m cột. dp[mask] = số cách khi mask là các ô "thừa ra" (bị thò sang) cột đang xét.
    dp = [0] * size
    dp[0] = 1  # trước cột đầu tiên: không có ô nào thò sang.
    for _ in range(m):
        dp = step(dp)

    # Sau m cột không được có ô nào thò ra ngoài lưới -> lấy dp[0].
    print(dp[0] % MOD)


main()

File: cses/03_dynamic_programming/2181_counting-tilings/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// Liệt kê mọi cách lát đầy một cột: với current_mask là các ô đã bị domino ngang
// từ cột trước thò sang chiếm chỗ, sinh mọi next_mask (tập ô thò sang cột kế).
void generate_next_masks(int row, int height, int current_mask, int next_mask,
                         vector<int>& transitions) {
    if (row == height) {
        transitions.push_back(next_mask);
        return;
    }

    // Ô đã bị chiếm sẵn từ cột trước -> bỏ qua, không thò sang cột kế.
    if ((current_mask & (1 << row)) != 0) {
        generate_next_masks(row + 1, height, current_mask, next_mask, transitions);
        return;
    }

    // Lựa chọn 1: đặt domino ngang thò sang cột kế -> bật bit row trong next_mask.
    generate_next_masks(row + 1, height, current_mask, next_mask | (1 << row), transitions);

    // Lựa chọn 2: đặt domino dọc phủ (row, row+1) ngay trong cột này (không thò sang).
    if (row + 1 < height && (current_mask & (1 << (row + 1))) == 0) {
        generate_next_masks(row + 2, height, current_mask, next_mask, transitions);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    constexpr int MOD = 1'000'000'007;

    int n, m;
    cin >> n >> m;
    // Đối xứng chuyển vị: lát (n x m) bằng lát (m x n). Chọn n nhỏ hơn để 2^n nhỏ nhất.
    if (n > m) {
        swap(n, m);
    }

    // Precompute các cạnh chuyển: transitions[mask] = danh sách next_mask sinh ra từ mask.
    const int states = 1 << n;
    vector<vector<int>> transitions(states);
    for (int mask = 0; mask < states; ++mask) {
        generate_next_masks(0, n, mask, 0, transitions[mask]);
    }

    // DP qua m cột. dp[mask] = số cách khi mask là tập ô bị thò sang cột đang xét.
    vector<int> dp(states, 0), next(states, 0);
    dp[0] = 1;  // trước cột đầu tiên: không có ô nào thò sang.
    for (int column = 0; column < m; ++column) {
        fill(next.begin(), next.end(), 0);
        for (int current_mask = 0; current_mask < states; ++current_mask) {
            if (dp[current_mask] == 0) {
                continue;
            }
            // Cộng dồn dp[current_mask] vào mọi next_mask hợp lệ (giữ trong khoảng [0, MOD)).
            for (const int next_mask : transitions[current_mask]) {
                next[next_mask] += dp[current_mask];
                if (next[next_mask] >= MOD) {
                    next[next_mask] -= MOD;
                }
            }
        }
        dp.swap(next);
    }

    // Sau m cột không được có ô nào thò ra ngoài lưới -> đáp án là dp[0].
    cout << dp[0] << '\n';
    return 0;
}
